rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,385 @@
+We can read container logs directly from the host filesystem. Kubernetes stores
+logs such that they're accessible from the following locations:
+
+- [`/var/log/pods`][var_log_pods_src];
+- `/var/log/containers` - legacy location, kept for backward compatibility
+  with pre `1.14` clusters.
+
+To make our lives easier, here's a [link][build_container_logs_directory_src] to
+the part of the k8s source that's responsible for building the path to the log
+file. If we encounter issues, this would be a good starting point to unwrap the
+k8s code.
+
+#### Log file format
+
+As already been mentioned above, log formats can vary, but there are certain
+invariants that are imposed on the container runtimes by the implementation of
+Kubernetes itself.
+
+A particularity interesting piece of code is the [`ReadLogs`][k8s_src_read_logs]
+function - it is responsible for reading container logs. We should carefully
+inspect it to gain knowledge on the edge cases. To achieve the best
+compatibility, we can base our log files consumption procedure on the logic
+implemented by that function.
+
+Based on the [`parseFuncs`][k8s_src_parse_funcs] (that
+[`ReadLogs`][k8s_src_read_logs] uses), it's evident that k8s supports the
+following formats:
+
+- Docker [JSON File logging driver] format - which is essentially a simple
+  [`JSONLines`][jsonlines] (aka `ndjson`) format;
+- [CRI format][cri_log_format].
+
+We have to support both formats.
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+1. [Fluentd k8s integration]
+1. [LogDNA k8s integration]
+1. [Honeycomb integration]
+1. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+1. [Influx Helm charts]
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community-driven
+   integrations.
+
+## Outstanding Questions
+
+### From Ben
+
+1. What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].
+1. What is the best to avoid Vector from ingesting it's own logs? I'm assuming
+   that my [`kubectl` tutorial](#kubectl-interface) handles this with namespaces?
+   We'd just need to configure Vector to exclude this namespace?
+1. I've seen two different installation strategies. For example, Fluentd offers
+   a [single daemonset configuration file][fluentd_daemonset] while Fluentbit
+   offers [four separate configuration files][fluentbit_installation]
+   (`service-account.yaml`, `role.yaml`, `role-binding.yaml`, `configmap.yaml`).
+   Which approach is better? Why are they different?
+1. Should we prefer `kubectl create ...` or `kubectl apply ...`? The examples
```

Comment:
We should support both. In our guides, we probably should use `kubectl apply`, since we specify a cluster-wide static resource.

`kubectl apply` requires some additional considerations compared to `kubectl create` to properly support idempotent upgrades.

Reply:
Explained this in details at the RFC itself.